kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,847 @@
+use crate::path::ParsedLogPath;
+use crate::scan::data_skipping::DataSkippingFilter;
+use crate::scan::scan_row_schema;
+use crate::scan::state::DvInfo;
+use crate::schema::{ArrayType, ColumnNamesAndTypes, DataType, MapType, SchemaRef, StructType};
+use crate::table_properties::TableProperties;
+use crate::utils::require;
+use crate::{
+    DeltaResult, Engine, EngineData, Error, ExpressionHandler, ExpressionRef, JsonHandler,
+    RowVisitor,
+};
+use itertools::Itertools;
+
+pub struct TableChangesScanData {
+    pub data: Box<dyn EngineData>,
+    pub selection_vector: Vec<bool>,
+    pub remove_dv: Option<Arc<HashMap<String, DvInfo>>>,
+}
+
+/// Given an iterator of ParsedLogPath and a predicate, returns an iterator of
+/// [`TableChangesScanData`]  Each row that is selected in the returned
+/// `engine_data` _must_ be processed to complete the scan. Non-selected rows
+/// _must_ be ignored.
+pub(crate) fn table_changes_action_iter(
+    engine: &dyn Engine,
+    commit_files: impl IntoIterator<Item = ParsedLogPath>,
+    table_schema: SchemaRef,
+    predicate: Option<ExpressionRef>,
+) -> DeltaResult<impl Iterator<Item = DeltaResult<TableChangesScanData>>> {
+    let json_handler = engine.get_json_handler();
+    let filter = DataSkippingFilter::new(engine, &table_schema, predicate);
+
+    let expression_handler = engine.get_expression_handler();
+    let result = commit_files
+        .into_iter()
+        .map(move |commit_file| -> DeltaResult<_> {
+            let mut scanner = LogReplayScanner::new(
+                commit_file,
+                json_handler.clone(),
+                filter.clone(),
+                expression_handler.clone(),
+                table_schema.clone(),
+            );
+            scanner.prepare_phase()?;
+            scanner.into_scan_batches()
+        })
+        .flatten_ok()
+        .map(|x| x?);
+    Ok(result)
+}
+
+// Gets the expression for generating the engine data in [`TableChangesScanData`].
+//
+// TODO: This expression is temporary. In the future it will also select `cdc` and `remove` actions
+// fields.
+fn get_add_transform_expr() -> Expression {
+    Expression::Struct(vec![
+        column_expr!("add.path"),
+        column_expr!("add.size"),
+        column_expr!("add.modificationTime"),
+        column_expr!("add.stats"),
+        column_expr!("add.deletionVector"),
+        Expression::Struct(vec![column_expr!("add.partitionValues")]),
+    ])
+}
+
+/// Processes a single commit file from the log to generate an iterator of [`TableChangesScanData`].
+/// The scanner operates in two phases that _must_ be performed in the following order:
+/// 1. Prepare phase [`LogReplayScanner::prepare_phase`]: This performs one iteration over every
+///    action in the commit. In this phase, we do the following:
+///     - Find the timestamp from a `CommitInfo` action if it exists. These are generated when
+///       In-commit timestamps is enabled.
```

Comment:
Do we actually support ICT yet in kernel? It changes how time travel resolution works, for example?

Reply:
Doesn't seem like we have it enabled, I don't see it in `TableProperties`. 
> It changes how time travel resolution works, for example?

Yeah anywhere there are timestamps, we need to use the ICT if it's enabled. This could be for Snapshot timetravel, or specifying a CDF with a timestamp. For CDF `_commit_timestamp` field, the [behaviour is to use ICT if it's there](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#reader-requirements-for-addcdcfile).

Reply:
Now updated to ignore ICT

Reply:
Note that we can't error out on ICT being present because it's not in TableProperties yet. @zachschuermann  this will probably go into the TableProperties/TableConfiguration update

Reply:
I have a PR up right now to add ICT as table property. Let's separate the two. I'd prefer to have a follow-up issue here that's specifically for 'enable ICT for CDF reads'